args/audio/systems/audiosystem.inl

```diff
@@ -0,0 +1,273 @@
+
+		m_sourcePositions.emplace(handle, event->entity.read_component<position>());
+
+		handle.write(a);
+		++sourceCount;
+	}
+
+
+	inline void AudioSystem::onAudioSourceComponentDestroy(events::component_destruction<audio_source>* event)
+	{
+		auto handle = event->entity.get_component_handle<audio_source>();
+		audio_source a = handle.read();
+
+		m_sourcePositions.erase(handle);
+
+		handle.write(a);
+		--sourceCount;
+	}
+
+	inline void AudioSystem::onAudioListenerComponentCreate(events::component_creation<audio_listener>* event)
+	{
+		log::debug("Creating Audio Listener...");
+
+		auto handle = event->entity.get_component_handle<audio_listener>();
+		audio_listener a = handle.read();
+		++listenerCount;
+		if (listenerCount > 1)
+		{
+			event->entity.remove_component<audio_listener>();
+		}
+		else
+		{
+			// listener count == 1
+			m_listenerEnt = event->entity;
+			setListener(event->entity.read_component<position>(), event->entity.read_component<rotation>());
+			m_listenerPosition = event->entity.read_component<position>();
+		}
+		handle.write(a);
+	}
+
+	inline void AudioSystem::onAudioListenerComponentDestroy(events::component_destruction<audio_listener>* event)
+	{
+		log::debug("Destroying Audio Listener...");
+
+		listenerCount = math::max((int)(listenerCount-1), 0);
+		if (listenerCount == 0)
+		{
+			log::debug("No Listeners left, resetting listener");
+			m_listenerEnt = ecs::entity_handle();
+			// Reset listener
+			async::readwrite_guard guard(m_contextLock);
+			alcMakeContextCurrent(alContext);
+			alListener3f(AL_POSITION, 0, 0, 0);
+			alListener3f(AL_VELOCITY, 0, 0, 0);
+			ALfloat ori[] = { 0, 0, 1.0f, 0, 1.0f, 0 };
+			alListenerfv(AL_ORIENTATION, ori);
+		}
+	}
+
+
+#pragma endregion
+
+	inline bool AudioSystem::initSource(audio_source& source)
+	{
+		async::readwrite_guard guard(m_contextLock);
+		alcMakeContextCurrent(alContext);
+
+		alGenSources((ALuint)1, &source.m_sourceId);
+		alSourcef(source.m_sourceId, AL_PITCH, 1);
+		alSourcef(source.m_sourceId, AL_GAIN, 1);
+		alSourcef(source.m_sourceId, AL_LOOPING, AL_TRUE);
+
+		// 3D audio stuffs
+		alSourcef(source.m_sourceId, AL_ROLLOFF_FACTOR, 2.0f);
+		alSourcef(source.m_sourceId, AL_REFERENCE_DISTANCE, 6);
+		alSourcef(source.m_sourceId, AL_MAX_DISTANCE, 15);
+
+		// NOTE TO SELF:
+		//		Set position and velocity to entity position and velocty
+		alSource3f(source.m_sourceId, AL_POSITION, 0, 0, 0);
+		alSource3f(source.m_sourceId, AL_VELOCITY, 0, 0, 0);
+
+		// Moved gen buffers to audio_importers.cpp
+		//alGenBuffers((ALuint)1, &source.m_audioBufferId);
+
+		auto [lock, segment] = source.m_audio_handle.get();
+
+		// Moved alBufferData to audio_importers.cpp
+		/*{
+			async::readonly_guard guard(lock);
+			alBufferData(source.m_audioBufferId, AL_FORMAT_MONO16, segment.buffer, segment.samples * sizeof(int16), segment.sampleRate);
+		}*/
+		
+		alSourcei(source.m_sourceId, AL_BUFFER, segment.audioBufferId);
+		alcMakeContextCurrent(nullptr);
+
+		return true;
```

Comment:
this can never fail? if so maybe remove the return bool and return void instead. right now this function seems very scary. it tells me it's failable, and if it fails the engine might just straight up quit... so every time I add an audio source there's a chance the engine will just fail and quit.